### PR TITLE
Controlling diagnostic display through function

### PR DESCRIPTION
**Pull request by ThePrimeagen**
I like the idea of having a function to override overall behavior as it is less cumbersome and more extensible as I get to control a lot of features of it.  It could even have a nil return to exclude the diagnostic notification.  

thoughts?

:wq

### CODE REVIEW

**Comment by ThePrimeagen** [comment]
oops, that second commit shouldn't of made it.  Showing stream a cool way to rebase, accidentally pushed bad commit.

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -674,7 +677,10 @@ function M.focusable_float(unique_name, fn)
       return
     end
   end
-  local pbufnr, pwinnr = fn()
+  local ok, pbufnr, pwinnr = pcall(fn())
```
Comment (norcalli) [inline runtime/lua/vim/lsp/util.lua:680]:
```suggestion
  local ok, pbufnr, pwinnr = pcall(fn)
```

**Comment by norcalli** [comment]
When I first wrote it, I assumed that people would take the lead in trying to take the code and refactor it into something more extensible, but that didn't happen. Considering some people have basically started copy pasting the entirety of the diagnostics code in order to make plugins like https://github.com/nvim-lua/diagnostic-nvim, I guess improving the ergonomics of how to customize the built-in stuff is probably a good idea.

Though this PR isn't nearly extensive enough for what I'm imagining.

**Comment by tjdevries** [comment]
@ThePrimeagen I think this might work for you? https://github.com/neovim/neovim/pull/12655/commits/57e658fef42cb893be12a030ec85d77f4140c2b7

Then you can do something like

```lua
vim.lsp.diagnostic.get_virtual_text_chunks_for_line = function(bufnr, line, line_diagnostics)
  ...
end
```

and override it.

**Comment by teto** [comment]
closing for cleanup. Feel free to reopen

**Comment by tjdevries** [comment]
This is now possible with new diagnostics, so it's all good :)

**Comment by cooper-anderson** [comment]
@tjdevries Do you have a clip going over "new diagnostics"?

**Comment by tjdevries** [comment]
@cooper-anderson yeah, a bunch are linked right here: https://github.com/neovim/neovim/pull/12655#issue-451693277

**Pull request closed without merging**